homeassistant/components/websocket_api/automation.py

```diff
@@ -0,0 +1,151 @@
+"""Automation related helper methods for the Websocket API."""
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+import logging
+from typing import Any, Self
+
+from homeassistant.const import CONF_TARGET
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import target as target_helpers
+from homeassistant.helpers.entity import (
+    entity_sources,
+    get_device_class,
+    get_supported_features,
+)
+from homeassistant.helpers.trigger import (
+    async_get_all_descriptions as async_get_all_trigger_descriptions,
+)
+from homeassistant.helpers.typing import ConfigType
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass(slots=True, kw_only=True)
+class _EntityFilter:
+    """Single entity filter configuration with AND logic for criteria."""
+
+    domains: set[str]
+    device_classes: set[str]
+    supported_features: set[int]
+
+    def matches(self, hass: HomeAssistant, entity_id: str, domain: str) -> bool:
+        """Return if entity matches all criteria in this filter."""
+        if self.domains and domain not in self.domains:
+            return False
+
+        if self.device_classes:
+            entity_device_class = get_device_class(hass, entity_id)
+            if (
+                entity_device_class is None
+                or entity_device_class not in self.device_classes
+            ):
+                return False
+
+        if self.supported_features:
+            entity_supported_features = get_supported_features(hass, entity_id)
+            if not any(
+                feature & entity_supported_features
+                for feature in self.supported_features
+            ):
+                return False
```

Comment:
I think the code is already correct, but please make sure to cover these cases in tests:

Filter requires both TURN_OFF both TURN_ON to be set:

```yaml
toggle:
  target:
    entity:
      domain: media_player
      supported_features:
        - - media_player.MediaPlayerEntityFeature.TURN_OFF
          - media_player.MediaPlayerEntityFeature.TURN_ON
```

Filter requires at least one of VOLUME_SET and VOLUME_STEP to be set:
volume_up:
  target:
    entity:
      domain: media_player
      supported_features:
        - media_player.MediaPlayerEntityFeature.VOLUME_SET
        - media_player.MediaPlayerEntityFeature.VOLUME_STEP
```

Reply:
Good catch. The first one is new to me and the code did not support it. Fixed!